File: android/src/main/kotlin/com/sunilflutter/flutter_ivs_stage/ParticipantData.kt
package com.sunilflutter.flutter_ivs_stage

import com.amazonaws.ivs.broadcast.ParticipantInfo
import com.amazonaws.ivs.broadcast.Stage
import com.amazonaws.ivs.broadcast.StageStream

class ParticipantData(
    val isLocal: Boolean,
    var participantId: String? = null,
    var attributes: Map<String, String> = emptyMap()
) {
    var publishState: Stage.PublishState = Stage.PublishState.NOT_PUBLISHED
    var subscribeState: Stage.SubscribeState = Stage.SubscribeState.NOT_SUBSCRIBED
    var streams: MutableList<StageStream> = mutableListOf()
    var wantsAudioOnly: Boolean = false
    var requiresAudioOnly: Boolean = false

    val isAudioOnly: Boolean
        get() = wantsAudioOnly || requiresAudioOnly

    val broadcastSlotName: String
        get() = if (isLocal) {
            "localUser"
        } else {
            val id = participantId
                ?: throw IllegalStateException("non-local participants must have a participantId")
            "participant-$id"
        }

    fun toMap(): Map<String, Any?> {
        val streamMaps = streams.map { stream ->
            mapOf(
                "deviceId" to (stream.device?.descriptor?.urn ?: ""),
                "type" to if (stream.streamType == StageStream.Type.VIDEO) "video" else "audio",
                "isMuted" to stream.muted
            )
        }
        return mapOf(
            "isLocal" to isLocal,
            "participantId" to participantId,
            "publishState" to publishState.toFlutterString(),
            "subscribeState" to subscribeState.toFlutterString(),
            "streams" to streamMaps,
            "wantsAudioOnly" to wantsAudioOnly,
            "requiresAudioOnly" to requiresAudioOnly,
            "broadcastSlotName" to broadcastSlotName,
            "attributes" to attributes
        )
    }
}

// Extension functions to match the iOS string format exactly
fun Stage.PublishState.toFlutterString(): String = when (this) {
    Stage.PublishState.NOT_PUBLISHED -> "notPublished"
    Stage.PublishState.ATTEMPTING_PUBLISH -> "attemptingPublish"
    Stage.PublishState.PUBLISHED -> "published"
}

fun Stage.SubscribeState.toFlutterString(): String = when (this) {
    Stage.SubscribeState.NOT_SUBSCRIBED -> "notSubscribed"
    Stage.SubscribeState.ATTEMPTING_SUBSCRIBE -> "attemptingSubscribe"
    Stage.SubscribeState.SUBSCRIBED -> "subscribed"
}

fun Stage.ConnectionState.toFlutterString(): String = when (this) {
    Stage.ConnectionState.DISCONNECTED -> "disconnected"
    Stage.ConnectionState.CONNECTING -> "connecting"
    Stage.ConnectionState.CONNECTED -> "connected"
}

File: ios/Classes/ParticipantData.swift
import Foundation
import AmazonIVSBroadcast

class ParticipantData {
    let isLocal: Bool
    var participantId: String?
    var publishState: IVSParticipantPublishState = .notPublished
    var subscribeState: IVSParticipantSubscribeState = .notSubscribed
    var streams: [IVSStageStream] = []
    var wantsAudioOnly = false
    var requiresAudioOnly = false
    var attributes: [String: String] = [:]

    var isAudioOnly: Bool {
        return wantsAudioOnly || requiresAudioOnly
    }

    var broadcastSlotName: String {
        if isLocal {
            return "localUser"
        } else {
            guard let participantId = participantId else {
                fatalError("non-local participants must have a participantId")
            }
            return "participant-\(participantId)"
        }
    }

    init(isLocal: Bool, participantId: String?, attributes: [String: String] = [:]) {
        self.isLocal = isLocal
        self.participantId = participantId
        self.attributes = attributes
    }
    
    func toMap() -> [String: Any] {
        let streamMaps = streams.map { stream -> [String: Any] in
            return [
                "deviceId": stream.device.descriptor().urn,
                "type": stream.device is IVSImageDevice ? "video" : "audio",
                "isMuted": stream.isMuted
            ]
        }
        
        return [
            "isLocal": isLocal,
            "participantId": participantId ?? NSNull(),
            "publishState": publishState.description,
            "subscribeState": subscribeState.description,
            "streams": streamMaps,
            "wantsAudioOnly": wantsAudioOnly,
            "requiresAudioOnly": requiresAudioOnly,
            "broadcastSlotName": broadcastSlotName,
            "attributes": attributes
        ]
    }
}
